File: Tutorial/Examples/Basic_routing/Consumer_C#/Consumer_2/Consumer_2.cs
using System;
using System.Text;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;

// Create connection.
var factory = new ConnectionFactory{ HostName = "localhost"};

// Instantiate the connection and one channel.
using var connection = factory.CreateConnection();
using var channel = connection.CreateModel();

// Declare an exchange called routing.
channel.ExchangeDeclare(exchange: "routing", type: ExchangeType.Direct);

// Declare a queue and bind it using the "square" and "both" keys.
var queueName = channel.QueueDeclare().QueueName;
channel.QueueBind(queue: queueName, exchange: "routing", routingKey: "square");
channel.QueueBind(queue: queueName, exchange: "routing", routingKey: "both");

// Declare consumer.
var consumer = new EventingBasicConsumer(channel);

// Subscribe to the Received event.
consumer.Received += (model, ea) =>{
    // Get the body of the message as byte array, decode to string and print it.
    var body = ea.Body.ToArray();
    var message = Encoding.UTF8.GetString(body);
    Console.WriteLine($"[Square consumer] Message received: {message}");
};

// Start consuming messages.
channel.BasicConsume(queue: queueName, autoAck: true, consumer: consumer);
Console.WriteLine("[Square consumer] Started consuming");

// Wait for a keypress to exit.
Console.ReadKey();

File: Tutorial/Examples/Basic_routing/Consumer_Python/Consumer_2.py
import pika
from pika.exchange_type import ExchangeType


def on_message_received(channe, method, properties, body):
    print(f'[Circle consumer] Received message: {body}')


# Set connection parameters. If connecting to a real server localhost should be replaced by the server´s address.
connection_parameters = pika.ConnectionParameters('localhost')

# Instantiate a connection unsing the connection_parameters previously defined
connection = pika.BlockingConnection(connection_parameters)

# Instantiate a channel
channel = connection.channel()

# Instantiate an exchange and queue.
channel.exchange_declare(exchange='routing', exchange_type=ExchangeType.direct)
queue = channel.queue_declare(queue='', exclusive=True)

# Bind the queue using the "circle" and "shape" keys.
channel.queue_bind(exchange='routing',
                   queue=queue.method.queue, routing_key='circle')
channel.queue_bind(exchange='routing',
                   queue=queue.method.queue, routing_key='shape')

# Start consuming messages
channel.basic_consume(queue=queue.method.queue, auto_ack=True,
                      on_message_callback=on_message_received)

print('[Circle consumer] Started consuming messages')

channel.start_consuming()
